rcl_action/include/rcl_action/action_server.h

```diff
@@ -0,0 +1,667 @@
+ * be checked by this function and therefore no deliberate error will occur.
+ * `ros_result` should be allocated and populated with the ROS result message.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] action_server handle to the action server that will send the result response
+ * \param[in] goal_handle a struct containing goal state information related to the response
+ * \param[in] terminal_state the final state of the goal
+ * \param[in] ros_result a struct containing the ROS result message
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs. */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_send_result_response(
+  const rcl_action_server_t * action_server,
+  const rcl_action_goal_handle_t * goal_handle,
+  const rcl_action_goal_state_t terminal_state,
+  void * ros_result);
+
+/// Take a pending cancel request using a rcl_action_server_t.
+/**
+ * This is a non-blocking call.
+ *
+ * The following cancel policy applies based on the goal ID and the timestamp
+ * contained in `goal_info`:
+ *
+ * - If the goal ID is zero and timestamp is zero, cancel all goals.
+ * - If the goal ID is zero and timestamp is not zero, cancel all goals accepted
+ *   at or before the timestamp.
+ * - If the goal ID is not zero and timestamp is zero, cancel the goal with the
+ *   given ID regardless of the time it was accepted.
+ * - If the goal ID is not zero and timestamp is not zero, cancel the goal with the
+ *   given ID and all goals accepted at or before the timestamp.
+
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * // TODO(jacobperron): Maybe we need to expose a request ID output for sending the response
+ * \param[in] action_server handle to the action server that will take the cancel request
+ * \param[out] goal_info a preallocated struct containing cancel information for one or
+ *   more goals.
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ACTION_SERVER_TAKE_FAILED` if take failed but no error occurred
+ *         in the middleware, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_take_cancel_request(
+  const rcl_action_server_t * action_server,
+  rcl_action_goal_info_t * goal_info);
```

Comment:
Once a client library has a cancel request, where does the logic for telling if a goal is part of a cancel request live? Is that duplicated in `rclpy` and `rclcpp`?